### DISCUSSION TITLE
what's the best way to keep only certain properties while serializing the object using jackson - should work for nested objects

### DISCUSSION THREAD

**Discussion by sidharthmittal** [FasterXML/jackson-databind#4706] [Q&A]
I want to keep only additionalFields in the response, and that property can be present in nested objects as well.
I am trying to use jackson, since I am already using jackson to serialize my object.

```
public class Policy {
    String name;
    String number;
    List<Participant> participants;
    List<Asset> assets;
    Map<String, String> additionalFields;
}

public class Participant {
    String name;
    String number;
    List<Asset> assets;
    Map<String, String> additionalFields;
}

public class Asset {
    String name;
    String number;
    Map<String, String> additionalFields;
}
```

now, I want to exclude all fields except additionalFields on all the objects in response. 

I have tried the below implementation, but this has 2 issues - 

1. removes everything and only keep additionalFields on Policy level.
2. I get the following error when I do not want to apply a filter, but have the classes annotated with filter - Cannot resolve PropertyFilter with id &#39;additionalFieldsFilter&#39;; no FilterProvider configured

`SimpleFilterProvider filterProvider = new SimpleFilterProvider() .addFilter("additionalFieldsFilter", SimpleBeanPropertyFilter.filterOutAllExcept("additionalFields"));`

Is there something simpler that I can use?

**Comment by pjfanning** [comment]
I'm sorry but there are lots of articles online about Jackson. We are volunteers and we simply can't answer every user question.

I would recommend https://www.baeldung.com/jackson-annotations - lots of annotations to try out. `@JsonIgnore` and `@JsonProperty` are good ones to try. You could also try writing your own custom serializer and you'll find Baeldung articles about this or there is lots to find using a search engine.

**Reply by sidharthmittal** [reply]
Thank you for the recommendation. I was wondering why this filter is not working on nested objects/list of nested object?

**Reply by cowtowncoder** [reply]
Jackson serializers are only responsible of serializing immediate descendants, and use delegation for others. As such all annotations by default only affect direct properties; they have no access or control over nested things as those are handled by different serializers. So implementing support for nested filtering would include adding some level of coordination across serializers.

An older blog post about various options is: http://www.cowtowncoder.com/blog/archives/2011/09/entry_461.html but it may not help much in your case.
